dvc/repo/artifacts.py

```diff
@@ -177,15 +201,28 @@ def download(
     ) -> tuple[int, str]:
         """Download the specified artifact."""
         from dvc.fs import download as fs_download
+        from dvc.repo import Repo
 
         logger.debug("Trying to download artifact '%s' via DVC", name)
         rev = self.get_rev(name, version=version, stage=stage)
+
+        dirname, _ = self.parse_path(name)
         with self.repo.switch(rev):
-            path = self.get_path(name)
+            root = self.repo.fs.root_marker
+            _dirname = self.repo.fs.join(root, dirname) if dirname else root
+            with Repo(_dirname, fs=self.repo.fs, scm=self.repo.scm) as r:
+                path = r.artifacts.get_path(name)
+                path = self.repo.fs.join(root, as_posix(path))
+                path = self.repo.fs.relpath(path, self.repo.root_dir)
+                # when the `repo` is a subrepo, the path `/subrepo/myart.pkl` for dvcfs
+                # should be translated as `/myart.pkl`,
+                # i.e. relative to the root of the subrepo
+                path = self.repo.fs.join(root, path)
+                path = self.repo.fs.normpath(path)
```

Comment:
Minor, but can we move all this inside `get_path()` so we don't have to duplicate it here and in `dvc.api.artifacts_show()`?

Reply:
We cannot move this to `get_path()`. `get_path()` returns git-root relative path.

Here, we are transforming the git-root relative path to well, `repo` relative path (the `repo` kwarg). In case of external repository, they are the same thing.

But for `repo=local_subrepo`, we have to make it relative to the `local_subrepo` instead of git-root relative path.

The _duplication_ bothers me too, but abstracting those out will make it more complicated (I am assuming the `repo.fs` is a posixpath as we are under `_repo.switch(rev) which is a `GitFileSystem`).

Also, both have a slightly different purpose, although they look almost the same. 
The purpose on the `dvc.api` is to translate to `os.path` (aka native path).
On `download()`, the purpose is to pass it to `dvcfs`.

Reply:
The path stuff is a mess, and very complicated to deal with when you have multiple paths in different format (`posixpath`, `os.path`, `git path`, `repo.fs` path, etc).